mgplvm/models/bfa.py

```diff
@@ -0,0 +1,296 @@
+# Bayesian Factor Analaysis
+import abc
+import torch
+import torch.nn as nn
+from torch import Tensor
+import numpy as np
+from mgplvm.utils import softplus
+from ..base import Module
+from ..kernels import Kernel
+from ..inducing_variables import InducingPoints
+from typing import Tuple, List, Optional, Union
+from torch.distributions import MultivariateNormal, LowRankMultivariateNormal, kl_divergence, transform_to, constraints, Normal
+from ..likelihoods import Likelihood
+
+jitter: float = 1E-8
+log2pi: float = np.log(2 * np.pi)
+
+
+class Bfa(Module):
+    """
+    Bayesian Factor Analysis
+    Assumes Gaussian observation noise
+    Computes log_prob and posterior predictions exactly
+    """
+
+    def __init__(self,
+                 n: int,
+                 sigma: Optional[Tensor] = None,
+                 learn_sigma=True):
+        super().__init__()
+        if sigma is None:
+            sigma = torch.ones(n,)  # TODO: FA init
+        self._sigma = nn.Parameter(data=sigma, requires_grad=learn_sigma)
+
+    @property
+    def prms(self) -> Tensor:
+        variance = torch.square(self._sigma)
+        return variance
+
+    @property
+    def sigma(self) -> Tensor:
+        return (1e-20 + self.prms).sqrt()
+
+    def log_prob(self, y, x):
+        m = x.shape[-1]
+        d = x.shape[-2]
+        n = y.shape[-2]
+        cov_factor = x[..., None, :, :].transpose(-1, -2)
+        cov_diag = self.prms[:, None] * torch.ones(m)
+        dist = LowRankMultivariateNormal(loc=torch.zeros(n, m),
+                                         cov_factor=cov_factor,
+                                         cov_diag=cov_diag)
+        lp = dist.log_prob(y)
+        return lp.sum()
+
+    def predict(self, xstar, y, x, full_cov=False):
+        m = x.shape[-1]
+        d = x.shape[-2]
+        n = y.shape[-2]
+        x = x[..., None, :, :]
+        cov_factor = x.transpose(-1, -2)
+        cov_diag = self.prms[:, None] * torch.ones(m)
+        dist = LowRankMultivariateNormal(loc=torch.zeros(n, m),
```

Comment:
might be worth having a separate method to construct this dist object to avoid duplicating code

Reply:
I've added this now